Fix node being undetected as part of function signature when too deeply nested.

niconoe- opened the pull request:
This PR:

- [x] Covered by tests
- [x] Fixes https://github.com/infection/infection/issues/1846

To detect if a Node is part of a signature of a function/method, it used to be checked if that Node is a "function-like" Node or if its direct parent is a Node\Param. 
With the usage of PHP Attributes, Nodes to check can be more deeply nested, and looking at the direct parent is no longer sufficient.
This PR ensures that all ancestors are being parsed until a Node\Param is found, or no more parent exists.

sanmai reviewed (approved):
The change to the visitor makes sense, but I didn't quite get the tests (probably because I'm too tired)

niconoe- commented:
> The change to the visitor makes sense, but I didn't quite get the tests (probably because I'm too tired)

I basically copy-pasted the integration test "ReflectionVisitorTest::test_it_marks_nodes_which_are_part_of_the_function_signature", but I added an attribute with a "Node\Expr\ConstFetch" element within in the duplicated test. As this node is actually part of the method signature, we expect `true`, but it cannot be detected as part of the method signature without my fix: the parent's node is not a "Node\Param" while it has an ancestor which is a "Node\Param".

In a clearer way, you can try with the Unit Test I added without the source code edition and you'll see that the dataset with the provider about the "Node\Expr\ConstFetch" node won't pass the test, while when applying my fix, it passes. :slightly_smiling_face:

maks-rafalko reviewed (approved):
Spent a couple of minutes with tests and yeah, they do make sense, as well as the source code change.

I absolutely appreciate your time debugging Infection and finding the root and fixing it, thank you.

maks-rafalko commented:
Regarding the differencies on Playground before and after update: I believe this is related to different versions of PHP (8.0 before, 8.1 currently), PHPUnit (9.5 before, 10.1 currently). Because too much has changed related to coverage.
